// hazelcast-client/lib/core/EntryEventType.js
"use strict";
/*
 * Copyright (c) 2008-2018, Hazelcast, Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
/* tslint:disable:no-bitwise */
var EntryEventType;
(function (EntryEventType) {
    EntryEventType[EntryEventType["ADDED"] = 1] = "ADDED";
    EntryEventType[EntryEventType["REMOVED"] = 2] = "REMOVED";
    EntryEventType[EntryEventType["UPDATED"] = 4] = "UPDATED";
    EntryEventType[EntryEventType["EVICTED"] = 8] = "EVICTED";
    EntryEventType[EntryEventType["EVICT_ALL"] = 16] = "EVICT_ALL";
    EntryEventType[EntryEventType["CLEAR_ALL"] = 32] = "CLEAR_ALL";
    EntryEventType[EntryEventType["MERGED"] = 64] = "MERGED";
    EntryEventType[EntryEventType["EXPIRED"] = 128] = "EXPIRED";
    EntryEventType[EntryEventType["INVALIDATION"] = 256] = "INVALIDATION";
})(EntryEventType = exports.EntryEventType || (exports.EntryEventType = {}));
