// hazelcast-client/lib/index.js
"use strict";
/*
 * Copyright (c) 2008-2018, Hazelcast, Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var HazelcastClient_1 = require("./HazelcastClient");
exports.Client = HazelcastClient_1.default;
var Config = require("./config/Config");
exports.Config = Config;
var ClientInfo_1 = require("./ClientInfo");
exports.ClientInfo = ClientInfo_1.ClientInfo;
var Predicates = require("./core/Predicate");
exports.Predicates = Predicates;
var Address = require("./Address");
exports.Address = Address;
var TopicOverloadPolicy = require("./proxy/topic/TopicOverloadPolicy");
exports.TopicOverloadPolicy = TopicOverloadPolicy;
var HazelcastErrors = require("./HazelcastError");
exports.HazelcastErrors = HazelcastErrors;
var Predicate_1 = require("./core/Predicate");
exports.IterationType = Predicate_1.IterationType;
var Aggregators = require("./aggregation/Aggregators");
exports.Aggregators = Aggregators;
